cap/modules/auth/views.py

```diff
@@ -0,0 +1,156 @@
+
+
+@blueprint.route('/authorize/<name>')
+def authorize(name):
+    if not current_user.is_authenticated:
+        return jsonify({"error": ANAUTHORIZED_MSG(name)}), 401
+
+    ui_flag = session.pop('ui', None)
+
+    client = current_auth.create_client(name)
+    token = client.authorize_access_token()
+
+    configs = OAUTH_SERVICES.get(name.upper(), {})
+    extra_data_method = configs.get('extra_data_method')
+
+    # TOFIX Add error handlers for reject, auth errors, etc
+    extra_data = {}
+    if (extra_data_method):
+        extra_data = extra_data_method(client, token)
+
+    _token = OAuth2Token.get(name=name, user_id=current_user.id)
+    if not _token:
+        _token = OAuth2Token(name=name, user_id=current_user.id)
+    _token.token_type = token.get('token_type', 'bearer')
+    _token.access_token = token.get('access_token')
+    _token.refresh_token = token.get('refresh_token')
+    _token.expires_at = token.get('expires_at', 0)
+    _token.expires_in = token.get('expires_in', 0)
+    _token.extra_data = extra_data
+    db.session.add(_token)
+
+    # Add extra data to user profile.
+    # If user profile doesn't exist yet, it creates one.
+    profile = UserProfile.get_by_userid(current_user.id)
+    if not profile:
+        profile = UserProfile(user_id=current_user.id)
+        db.session.add(profile)
+
+    profile_data = get_oauth_profile(name)
+
+    profile_services = profile.extra_data.get("services", {})
+    profile_services[name] = profile_data
+    profile.extra_data = {"services": profile_services}
+    flag_modified(profile, "extra_data")
+
+    db.session.commit()
+
+    if ui_flag:
+        if current_app.config['DEBUG']:
+            redirect_url = "http://localhost:3000/settings/auth/connect"
+        else:
+            redirect_url = "/settings/auth/connect"
+        return redirect(redirect_url)
+    else:
+        return jsonify({
+            "message": "Authorization to {} succeeded".format(name)
+        }), 200
+
+
+@blueprint.route('/profile/<name>')
+def profile(name):
+    if not current_user.is_authenticated:
+        return jsonify({"error": ANAUTHORIZED_MSG(name)}), 403
+
+    _token = OAuth2Token.get(name=name, user_id=current_user.id)
+    if not _token:
+        return jsonify({"message":
+                        "Your account is not connected to the service"}), 403
+
+    extra_data = _token.extra_data
+    client = current_auth.create_client(name)
+
+    if name == "github":
+        resp = client.get("/user")
+    elif name == "zenodo":
+        resp = client.get("api/")
+    elif name == "cern":
+        resp = client.get("Me")
+    elif name == "orcid":
+        orcid_id = extra_data.get('orcid_id')
+        if orcid_id:
+            resp = client.get("/{}/record".format(orcid_id),
+                              headers={'Accept': 'application/json'})
+
+    if resp:
+        profile = resp.json()
+
+    return jsonify(profile)
+
+
+def get_oauth_profile(name):
```

Comment:
you call it only after you have token already, pass it instead of querying again